### ISSUE TITLE
Linked monitors not recording when parent goes into alert

### ISSUE THREAD

**Issue by maddios** [ZoneMinder/zoneminder#3393]
**Describe Your Environment**
- Version of ZoneMinder: 1.36.10
- How you installed ZoneMinder: from-source
- Full name and version of OS: Ubuntu 18.04 LTS
- Server is a 16 core 2.4ghz xeon with 64GB ram, my whole setup runs at about 10 load average while using approx 30GB ram.

I'm running 5 Amcrest IP cameras each outputting 2 streams (low and high rez), I'm actually doing my triggering in reverse though, I have my high res streams running MoDect and the low res streams actually go to a ML object detection/motion tracking software I wrote which then outputs a stream which I'm also streaming into zoneminder, These 5 AI streams are as NoDect. Each of these 5 streams has its full res stream set as a Linked monitor.

**Describe the bug**
The issue I'm having is that MOST of the time my linked monitors don't record anything, I've tried restarting ZM and triggering an alert once ZM is back up and it'll sometimes trigger and sometimes not. As far as I can tell there's nothing in the log files to indicate any kind of issue even with debug mode enabled. I've check the zmc_m{id}.log files and the zmwatch.log files.

**To Reproduce**
I have no idea how to reproduce this other than it happens in my env all the time.

**Expected behavior**
I expect that every single time an alert is triggered all linked cameras must record it

**Debug Logs**
[parent](https://drive.google.com/file/d/18sYYX4M0xG34XrY8AT8-soHJauP1gtZH/view?usp=sharing)
[linked](https://drive.google.com/file/d/1LR9KLaedkIh8wzA_IwIR96W4LELh0iAo/view?usp=sharing)

first log shows an alert around 19:14:59.339395, but the 2nd one never does

**Comment by maddios** [comment]
Actually, I forgot to mention that the triggering monitor has a total of 4 linked monitors, 3 other cameras link to it as well as the AI stream. Not sure if this causes some sort of mutex issue

**Comment by connortechnology** [comment]
Whoops, accidentally closed this with an unrelated commit.

Um there are no mutexes involved.  we will need debug level 3 logs from all related zmc processes.

**Comment by maddios** [comment]
the logs I included were level 9, is level 3 more verbose? or does level 3 specifically log the mutexes?

**Comment by SteveGilvarry** [comment]
Only got INF records in the logs you posted, should see DBG lines in debug logs. Also normally set to log to separate files for each process. Level 3 is lower and less verbose than 9. So you might need it for both monitors using _zmc_mX options

This might help or just ask if no luck.
https://zoneminder.readthedocs.io/en/stable/userguide/options/options_logging.html

**Comment by maddios** [comment]
Wait where are you seeing INF records in these, these are the zm_debug.log.{id} files. When I download the 2 files I just get a bunch of stuff like this:

```
12/01/21 19:14:49.607387 zmc_m2[37395].DB4-zm_packetqueue.cpp/431 [Locking in get_packet using it 0x560ef41db6c0 queue end? 0]
12/01/21 19:14:49.607392 zmc_m2[37395].DB4-zm_packetqueue.cpp/436 [Have Lock in get_packet]
12/01/21 19:14:49.607397 zmc_m2[37395].DB3-zm_packetqueue.cpp/450 [get_packet using it 0x560ef41db6c0 locking index 33]
12/01/21 19:14:49.607402 zmc_m2[37395].DB4-zm_packet.h/106 [TryLocking packet 33]
```

But sure, I'll get a level 3 debug snapshot when I get a chance, i can see how level 9 might be too verbose.

**Comment by SteveGilvarry** [comment]
Sorry couldn't see the files from mobile app for GitHub. Was talking about the Debug logs you pasted into issue log.

**Comment by maddios** [comment]
Here are new zm debug logs for the 2 monitors:

[zm_debug_parent.log](https://github.com/ZoneMinder/zoneminder/files/7696226/zm_debug_parent.log)
[zm_debug_linked.log](https://github.com/ZoneMinder/zoneminder/files/7696227/zm_debug_linked.log)

**Comment by maddios** [comment]
is this a relevant thing? 

12/10/21 20:37:17.906460 zmc_m2[5368].DB1-zm_monitor.cpp/2336 [Reloading linked monitors for monitor front door, '(null)']

shouldn't null here be the 4 linked monitors?

**Comment by connortechnology** [comment]
So the issue here is that the analysis thread was waiting on capture instead of looking for a linked monitor to be triggered.
Capture took .2 of a second which is a long time actually.
So linked monitor checking needs to be  it's own thread...  I can't imagine making this kind of change in 1.36.  Maybe with enough testing.  
That being said... the linked monitor DID notice the alarmed state of the primary...  so I'm gonna have to look closer at the code.
Keep to that code says I,
Yeargh!

**Comment by maddios** [comment]
Thanks, I'm glad you can see the issue, it's ok if you can't do it in .36. I've been living with this "bug" for years, I just never really invested a ton of time into debugging it until now. Hopefully it can be addressed at some point in the future.

If you do release a fix I'll gladly test it.

**Comment by connortechnology** [comment]
I think I can improve it a lot.  Ultimate fix in a few months. At the end of the day we will always be polling and will have to sleep unless we consume 100% cpu.  
I think long term we will use mqtt to queue a notice about the alarm status and so even though we might notice it late, we will still get it.

**Comment by maddios** [comment]
yeah, even if it triggers within a second or two it should still be fine, at least for my setup. Can always increase the pre-alarm frame count.

**Comment by SteveGilvarry** [comment]
Went through the level-3 logs (parent = monitor 2 "front door", linked = monitor 7 "front door small"). The linked monitor is **not missing the alarm** — it detects it but never escalates to a recorded event.

Timeline:
- `20:37:42.198` — monitor 2 **Gone into alarm** (Forced Web), opens event 701381.
- `20:37:42.286` — monitor 7 sees "Linked monitor 2 front door **is alarmed**" (88 ms later) and **Gone into prealarm state**.
- `20:37:42.2`–`43.7` — monitor 2's forced alarm lasts ~1.5 s; monitor 7 sees it alarmed for **8 consecutive frames** and stays in prealarm.
- alarm clears → monitor 7 drops out of prealarm. There is **no "Opening new event" anywhere in monitor 7's log** — it never recorded.

Why it never escalated, two compounding causes:

1. **The linked monitor's analysis is badly backed up.** 100+ `Waiting for decode`, packet queue pinned at **46** deep, analysis advancing only ~**5 fps** (image_count 116→117→118→119 at ~200 ms apart) instead of realtime. Decode can't keep up with capture on a loaded box.
2. **AlarmFrameCount gating.** Escalating prealarm→alarm needs `AlarmFrameCount` consecutive alarmed frames. Monitor 2's is 10 (logged: `PreAlarmCount: 9 > AlarmFrameCount:10`). At ~5 fps the ~1.5 s alarm only yields **8** alarmed frames on monitor 7 — short of the threshold — so it never crosses into alarm and a non-escalated prealarm produces no event.

So linked **detection works**; the failure is that a linked alarm flows through the linked monitor's own `AlarmFrameCount` prealarm gating, which a backed-up monitor can't satisfy during a short parent alarm.

Workaround today: set the linked monitor's **AlarmFrameCount = 1**. That only suits the pure "low-res detects → high-res records" setup, though — here monitor 7 also runs its own motion detection, so `AlarmFrameCount = 1` would make its own motion hair-trigger, so it isn't a general fix.

Proposed fix: **a linked alarm should bypass `AlarmFrameCount` and escalate immediately**, the same way a forced/external trigger does, while the monitor's own motion detection keeps its `AlarmFrameCount`. That makes "every linked camera records the alert" reliable regardless of analysis backlog, without over-triggering the monitor's own motion.

Worth noting: the post-2021 change that set the score in shared memory before state transitions (`dc8f1bc8e`) improves *detection timing*, but detection isn't the problem in these logs, so it wouldn't fix this case.

@maddios — long shot given how long it's been, but if you still run this setup: what is the linked monitor's `AlarmFrameCount` set to? (Expecting the default 10.) The backed-up decode is also a factor — hardware decode / lower analysis FPS would widen the effective window.
